pkg/cloudprovider/provider/kubevirt/provider.go

```diff
@@ -651,13 +709,34 @@ func (p *provider) newVirtualMachine(ctx context.Context, c *Config, pc *provide
 					TopologySpreadConstraints:     getTopologySpreadConstraints(c, map[string]string{machineDeploymentLabelKey: labels[machineDeploymentLabelKey]}),
 				},
 			},
-			DataVolumeTemplates: getDataVolumeTemplates(c, dataVolumeName),
+			DataVolumeTemplates: getDataVolumeTemplates(c, dataVolumeName, getDataVolumeSource(c, mdName)),
 		},
 	}
 	return virtualMachine, nil
 }
 
-func (p *provider) Cleanup(ctx context.Context, machine *clusterv1alpha1.Machine, _ *cloudprovidertypes.ProviderData) (bool, error) {
+func (p *provider) createPreAllocatedDataVolume(ctx context.Context, c *Config, machine *clusterv1alpha1.Machine, sigClient client.Client, dvName string) error {
+	existingDV := &cdiv1beta1.DataVolume{}
+	exist := true
+	if err := sigClient.Get(ctx, types.NamespacedName{Namespace: c.Namespace, Name: dvName}, existingDV); err != nil {
+		if !kerrors.IsNotFound(err) {
+			return err
+		}
+		exist = false
+	}
+
+	// create preAllocated DataVolume, if it doesn't exist.
+	//  preAllocated DataVolume will be created per MachineDeployment.
+	if !exist {
+		return sigClient.Create(ctx, newDataVolume(c, dvName, machine.Name))
+	}
+	// annotate existing DataVolume with VM-name, if already exist.
+	updatedDV := existingDV.DeepCopy()
```

Comment:
any reason for doing a `DeepCopy` in here? Controller runtime get function will return already a deep copy.